File: performance-benchmark/src/test/scala/net/ssimmie/todos/performance/benchmark/BenchmarkSimulation.scala
package net.ssimmie.todos.performance.benchmark

import io.gatling.core.Predef._
import io.gatling.core.structure.ScenarioBuilder
import io.gatling.http.Predef._
import io.gatling.http.protocol.HttpProtocolBuilder
import net.ssimmie.todos.performance.benchmark.ChecklistResourceAction.getChecklist
import net.ssimmie.todos.performance.benchmark.ChecklistsResourceAction.createChecklist
import net.ssimmie.todos.performance.benchmark.HealthCheckAction.healthcheck
import net.ssimmie.todos.performance.benchmark.RootResourceAction.queryRoot

import scala.concurrent.duration._

class BenchmarkSimulation extends Simulation {

  val httpProtocol: HttpProtocolBuilder = http
    .warmUp("http://localhost:8181/actuator")
    .baseUrl("http://localhost:8181")
    .acceptHeader("application/json")
    .shareConnections

  val healthCheckJourney: ScenarioBuilder = scenario("Todos Health Check")
    .exec(healthcheck)

  val coreJourney: ScenarioBuilder = scenario("Todos Core Journey")
    .exec(queryRoot, createChecklist, getChecklist)

  setUp(
    healthCheckJourney.inject(constantUsersPerSec(1).during(1.minutes)),
    coreJourney.inject(constantUsersPerSec(30).during(1.minutes).randomized)
  ).protocols(
    httpProtocol
  ).assertions(
    global.responseTime.max.lt(1000),
    global.responseTime.percentile4.lt(500),
    global.responseTime.percentile3.lte(125),
    global.successfulRequests.percent.is(100)
  )
}

File: performance-benchmark/src/test/java/net/ssimmie/todos/performance/benchmark/BenchmarkSimulation.java
package net.ssimmie.todos.performance.benchmark;

import io.gatling.javaapi.core.ScenarioBuilder;
import io.gatling.javaapi.core.Simulation;
import io.gatling.javaapi.http.HttpProtocolBuilder;

import java.time.Duration;

import static io.gatling.javaapi.core.CoreDsl.*;
import static io.gatling.javaapi.http.HttpDsl.*;

public class BenchmarkSimulation extends Simulation {

    HttpProtocolBuilder httpProtocol = http
            .warmUp("http://localhost:8181/actuator")
            .baseUrl("http://localhost:8181")
            .acceptHeader("application/json")
            .shareConnections();

    ScenarioBuilder healthCheckJourney = scenario("Todos Health Check")
            .exec(HealthCheckAction.healthcheck());

    ScenarioBuilder coreJourney = scenario("Todos Core Journey")
            .exec(
                    RootResourceAction.queryRoot(),
                    ChecklistsResourceAction.createChecklist(),
                    ChecklistResourceAction.getChecklist()
            );

    {
        setUp(
                healthCheckJourney.injectOpen(constantUsersPerSec(1).during(Duration.ofMinutes(1))),
                coreJourney.injectOpen(constantUsersPerSec(30).during(Duration.ofMinutes(1)).randomized())
        ).protocols(httpProtocol)
                .assertions(
                        global().responseTime().max().lt(1500),
                        global().responseTime().percentile4().lt(900),
                        global().responseTime().percentile3().lte(250),
                        global().successfulRequests().percent().is(100.0)
                );
    }
}